### PR TITLE
Helm Chart: Adding support for secrets enabled multi-cluster

### PR DESCRIPTION
**Pull request by cmcgalliard**
## Summary
This simplifies the process of configuring kubeconfigs in secrets for Headlamp. This functionality already exists; the Helm chart just doesn't make it easy to configure.

## Changes

- Updated Deployment YAML to add a volume/volumeMount for each kubeconfig secret
- Updated Deployment YAML to pass the joined kubeconfig paths to Headlamp via the `-kubeconfig` argument
- Updated values file to allow for a list of secrets containing kubeconfigs
- Updated helm documentation

## Steps to Test

1. Create 1+ secret containing a kubeconfig

2. Configure the `kubeconfigSecrets`
``` YAML
config:
  kubeconfigSecrets:
    - secretName: prod-cluster-kubeconfig
      key: config  # optional, defaults to "config"
```
3.  `helm install headlamp`
4.  Navigate to clusters and validate that you can see multiple clusters configured with the secrets you provided.

> Note: There is no need to disable `config.inCluster`. Because the chart now passes the kubeconfig paths via the `-kubeconfig` argument (instead of the `KUBECONFIG` env var, which is ignored in in-cluster mode), the in-cluster context and the clusters from the mounted kubeconfig secrets all load together into the same store.

## Screenshots (if applicable)
image

## Notes for the Reviewer

Reworked per review feedback to pass the kubeconfig paths via the `-kubeconfig` argument rather than the `KUBECONFIG` env var, so in-cluster and kubeconfig-secret contexts are no longer mutually exclusive.

### CODE REVIEW

File: charts/headlamp/README.md [hunk 1]
```diff
@@ -127,6 +127,34 @@ config:
...
+Headlamp supports managing multiple Kubernetes clusters by mounting kubeconfig files from secrets. This is ideal for fleet management scenarios where you want to access multiple clusters from a single Headlamp installation.
+
+#### Using Kubeconfig Secrets
+
+| Key | Type | Default | Description |
+|-----|------|---------|-------------|
+| config.kubeconfigSecrets | list | `[]` | List of secrets containing kubeconfig files |
+| config.kubeconfigSecrets[].secretName | string | required | Name of the secret containing kubeconfig |
+| config.kubeconfigSecrets[].key | string | `"config"` | Key within the secret data (optional, defaults to "config") |
+
```
[the remark this answers is not in the document]
Comment (cmcgalliard) [reply charts/headlamp/README.md:223]: I don't see a line where this is the case

File: charts/headlamp/values.yaml [hunk 2]
```diff
@@ -118,6 +118,19 @@ config:
...
+  # Allows mounting multiple Kubernetes secrets containing kubeconfig files
+  # IMPORTANT: When using kubeconfigSecrets, you must set config.inCluster to false
+  # The in-cluster mode and external kubeconfig loading are mutually exclusive
+  # Example:
+  # kubeconfigSecrets:
+  #   - secretName: prod-cluster-kubeconfig
+  #     key: config  # optional, defaults to "config"
+  #   - secretName: dev-cluster-kubeconfig
+  #     key: kubeconfig
+  #   - secretName: staging-cluster-kubeconfig
```
[the remark this answers is not in the document]
Comment (cmcgalliard) [reply charts/headlamp/values.yaml:191]: Adding `config:` would be more confusing than helpful, as it would add a duplicate line if someone were to just uncomment the code.

File: charts/headlamp/templates/deployment.yaml [hunk 3]
```diff
@@ -345,12 +355,19 @@ spec:
...
             - name: plugins-dir
               mountPath: {{ .Values.config.pluginsDir }}
             {{- end }}
+            {{- if include "headlamp.hasKubeconfigSecrets" . }}
+            {{- range $index, $secret := .Values.config.kubeconfigSecrets }}
+            - name: kubeconfig-{{ $index }}
+              mountPath: /kubeconfigs/{{ $secret.secretName }}
+              readOnly: true
+            {{- end }}
+            {{- end }}
```
[the remark this answers is not in the document]
Comment (cmcgalliard) [reply charts/headlamp/templates/deployment.yaml:450]: added a test accordingly

File: charts/headlamp/templates/deployment.yaml [hunk 4]
```diff
@@ -99,18 +99,28 @@ spec:
...
+            {{- if .Values.env }}
             {{- toYaml .Values.env | nindent 12 }}
+            {{- end }}
           {{- end }}
           {{- else }}
           env:
+            {{- if include "headlamp.hasKubeconfigSecrets" . }}
+            - name: KUBECONFIG
+              value: {{ include "headlamp.kubeconfigPath" . | quote }}
+            {{- end }}
```
[the remark this answers is not in the document]
Comment (cmcgalliard) [reply charts/headlamp/templates/deployment.yaml]: I am still not sure whether the lack of the `in-cluster` kubeconfig mapping here was intentional.  I also asked about this in the CNCF Slack. It would be nice if someone from the project could let me know what you all would like :)

**Comment by cmcgalliard** [comment]
Tested this evening after adjusting for code review and it seems to work as expected.

File: charts/headlamp/templates/deployment.yaml [hunk 5]
```diff
@@ -99,18 +99,28 @@ spec:
...
               name: {{ $oidc.externalSecret.name }}
-          {{- if .Values.env }}
+          {{- if or .Values.env (include "headlamp.hasKubeconfigSecrets" .) }}
           env:
+            {{- if include "headlamp.hasKubeconfigSecrets" . }}
+            - name: KUBECONFIG
+              value: {{ include "headlamp.kubeconfigPath" . | quote }}
+            {{- end }}
+            {{- if .Values.env }}
             {{- toYaml .Values.env | nindent 12 }}
```
[the remark this answers is not in the document]
Comment (cmcgalliard) [reply charts/headlamp/templates/deployment.yaml]: I think handling this edge case will make the helm difficult to read - most folks will not use both if they get this far, IMO.

**Comment by illume** [comment]
@cmcgalliard are you able to squash some of the commits?

**Comment by cmcgalliard** [comment]
done!

File: charts/headlamp/README.md [hunk 6]
```diff
@@ -127,6 +127,34 @@ config:
...
+      key: config
+    - secretName: dev-cluster-kubeconfig
+      key: kubeconfig
+    - secretName: staging-cluster-kubeconfig
+```
+#### Important Limitations
+
+**In-Cluster Mode and Kubeconfig Secrets are Mutually Exclusive**
+
+When `config.inCluster: true` (the default), Headlamp's backend explicitly skips loading external kubeconfig files. This is by design in the Headlamp backend code. If you want to use kubeconfig secrets to manage multiple clusters, you must set `config.inCluster: false`. This allows Headlamp to load the specified kubeconfig secrets and provide access to those clusters.
```
Comment (sniok) [inline charts/headlamp/README.md]: I don't think we can do this, when deploying helm chart to the cluster `inCluster` has to be true, it affects other logic and it's how we determine if headlamp is running in cluster/chart. 

The fact that KUBECONFIG env variable is ignored might be something we need to change because I don't think why we should ignore it in cluster.

As a workaround instead of setting a KUBECONFIG env variable we could pass a `-kubeconfig` argument instead, which shouldn't be ignored and should load multiple configs. @cmcgalliard

Comment (cmcgalliard) [reply charts/headlamp/README.md]: I'm not sure what else the `inCluster` option does, but it "works" on my cluster, pointing to other clusters with this branch, as shown in my screenshot above. 

Ideally, what do you suggest we do here? I can either dig into why `inCluster` fails to load the additional kubeconfigs in this PR. Or would you want sperate PR's targeting each change :) 

I will be on PTO this week, so my responses may be delayed.

**Comment by ashu8912** [comment]
Hii @cmcgalliard Thanks for the PR, the CI is failing and I think this needs a rebase

File: charts/headlamp/README.md (hunk 6 above)
Comment (sniok) [reply charts/headlamp/README.md:156]: I think passing kubeconfig path as a `-kubeoncfig` argument instead of the env variable would be the simplest solution here

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for these changes.


Can you please address the open review comments?

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


would you mind rebasing this branch on top of the latest main rather than merging it in?

<details>
<summary>Why this matters</summary>

Merge commits from `main` make the PR history harder to review. Please rebase your branch on top of the latest `main` instead, then update the PR with the rebased commits.

</details>

**Comment by cmcgalliard** [comment]
Sorry - life happened a bit
rebased & conflicts dealt with

**Comment by cmcgalliard** [comment]
Rebased onto latest upstream `main` and reworked the approach based on review feedback.

### What changed
Instead of exporting `KUBECONFIG` as an env var, the chart now passes the joined kubeconfig paths to Headlamp via the `-kubeconfig` argument:

```
- "-in-cluster"
- "-kubeconfig=/kubeconfigs/prod-cluster-kubeconfig/kubeconfig:/kubeconfigs/dev-cluster-kubeconfig/kubeconfig:/kubeconfigs/staging-cluster-kubeconfig/config"
```

The `KUBECONFIG` env var is ignored in in-cluster mode because `config.setKubeConfigPath` returns early before reading the env. Passing the same paths as `-kubeconfig` avoids that early return, so **in-cluster and kubeconfig-secret contexts now load together into the same store** — they are no longer mutually exclusive. No backend change required.

Also updated: the `multi-cluster` test case now sets `inCluster: true` to exercise the combined case, and the README/values comments drop the "mutually exclusive / must set inCluster: false" note.

### CI notes
- The changed CI workflows are `helm-chart-template-test.yml` (`make helm-template-test`) and `helm-chart-lint-test.yml` (`ct lint` + `ct install`).
- Ran the template test suite locally (`charts/headlamp/tests/test.sh`, the same thing `make helm-template-test` invokes): **all 45 template checks + failure-cases pass**, including the updated `multi-cluster` case.
- Ran `helm lint ./charts/headlamp --values tests/test_cases/multi-cluster.yaml`: **0 charts failed**.
- `.github/ct.yaml` sets `check-version-increment: false`, so no `Chart.yaml` version bump is needed.
- `ct install` (Minikube-based) runs in CI only; not reproduced locally.
- Note: EasyCLA re-evaluates on force-push and may show pending briefly.

Regenerated `tests/expected_templates/multi-cluster.yaml` to match the new rendered output.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


The [EasyCLA bot](https://github.com/kubernetes-sigs/headlamp/pull/4938#issuecomment-4107099744) is reporting that the CLA has not been signed. Kubernetes requires a signed CLA — could you please follow the link in that comment to sign it?

<details>
<summary>About the CLA</summary>

The Kubernetes project uses the Linux Foundation EasyCLA system. Signing takes only a minute through the link in the EasyCLA bot comment. Individual contributors sign electronically; corporate contributors may need their employer to sign a Corporate CLA.

</details>

**Comment by vtushar06** [comment]
@cmcgalliard @illume took a look. The wiring lines up: each secret mounts at /`kubeconfigs/<secretName>` with the key at path `<key>`, and the `-kubeconfig `arg builds` /kubeconfigs/<secretName>/<key>`, so the paths match. Switching to the arg also drops the old duplicate-KUBECONFIG-env concern, so that's two birds.

One edge case worth handling before you test: the mountPath is keyed on secretName (`/kubeconfigs/{{ $secret.secretName }})`, so if someone lists two entries pointing at the same secret (say one secret holding a couple of kubeconfig keys), both volumes mount at the same path and collide. Keying the mountPath on $index instead (e.g. `/kubeconfigs/{{ $index }})` and matching that in the kubeconfigPath helper would make it robust to that.

When you spin up clusters, the main thing to confirm is the combined case, the in-cluster context plus the secret clusters all showing together, since that's what this now enables.

File: charts/headlamp/templates/deployment.yaml [hunk 7]
```diff
@@ -553,6 +563,16 @@ spec:
...
           emptyDir: {}
         {{- end }}
+        {{- if include "headlamp.hasKubeconfigSecrets" . }}
+        {{- range $index, $secret := .Values.config.kubeconfigSecrets }}
+        - name: kubeconfig-{{ $index }}
+          secret:
+            secretName: {{ required "config.kubeconfigSecrets[].secretName is required" $secret.secretName }}
+            items:
+            - key: {{ default "config" $secret.key }}
+              path: {{ default "config" $secret.key }}
```
[the remark this answers is not in the document]
Comment (illume) [reply charts/headlamp/templates/deployment.yaml]: Yeah, this one is valid.

File: charts/headlamp/tests/expected_templates/multi-cluster.yaml [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            items:
+            - key: kubeconfig
+              path: kubeconfig
+        - name: kubeconfig-1
+          secret:
+            secretName: dev-cluster-kubeconfig
+            items:
+            - key: kubeconfig
+              path: kubeconfig
+        - name: kubeconfig-2
```
[the remark this answers is not in the document]
Comment (illume) [reply charts/headlamp/tests/expected_templates/multi-cluster.yaml:167]: this will be fixed when the deployment.yaml is updated

File: charts/headlamp/tests/expected_templates/multi-cluster.yaml [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+              mountPath: /kubeconfigs/2
+              readOnly: true
+      volumes:
+        - name: kubeconfig-0
+          secret:
+            secretName: prod-cluster-kubeconfig
+            items:
+            - key: kubeconfig
+              path: kubeconfig
+        - name: kubeconfig-1
```
[the remark this answers is not in the document]
Comment (illume) [reply charts/headlamp/tests/expected_templates/multi-cluster.yaml:161]: this will be fixed when the deployment.yaml is updated

**Comment by vtushar06** [comment]
@illume i ran the chart tests, the multi-cluster snapshot test and helm lint pass and the rendered output is valid YAML (items indentation looks right). Index-based paths line up with the mounts and the combined in-cluster + secrets case renders as expected. LGTM from my side 🚀🚀

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for the contribution.


There are some open Copilot review comments — could you take a look at them? Please mark each one as resolved once you've addressed it.

File: charts/headlamp/templates/deployment.yaml [hunk 10]
```diff
@@ -553,6 +563,16 @@ spec:
         - name: headlamp-tmp
           emptyDir: {}
         {{- end }}
+        {{- if include "headlamp.hasKubeconfigSecrets" . }}
+        {{- range $index, $secret := .Values.config.kubeconfigSecrets }}
+        - name: kubeconfig-{{ $index }}
+          secret:
+            secretName: {{ required "config.kubeconfigSecrets[].secretName is required" $secret.secretName }}
+            items:
```
[the remark this answers is not in the document]
Comment (cmcgalliard) [reply charts/headlamp/templates/deployment.yaml:571]: Good call — the error message now includes the entry index: `config.kubeconfigSecrets[%d].secretName is required` (via `printf` with `$index`), so it's clear which list entry is missing a `secretName`.

File: charts/headlamp/README.md [hunk 11]
```diff
@@ -202,49 +202,41 @@ config:
...
-plugins. Each entry renders as a Kubernetes `image` volume and is mounted
-read-only into the Headlamp container. If an access provider `execConfig.command`
-is configured, the command must be under one of the absolute
-`plugins[].mountPath` values.
+When `config.kubeconfigSecrets` is set, the chart mounts each secret and passes
+the joined kubeconfig paths to Headlamp via the `-kubeconfig` argument. This
+works alongside `config.inCluster` (the default), so the in-cluster context and
+the clusters from the mounted kubeconfig secrets are all loaded together into the
+same store. There is no need to disable `config.inCluster` to use kubeconfig
+secrets.
```
[the remark this answers is not in the document]
Comment (cmcgalliard) [reply charts/headlamp/README.md]: Reconciled — the README is correct: in-cluster and kubeconfig secrets now load together (we pass `-kubeconfig` instead of the ignored `KUBECONFIG` env var). Updated the PR description so the Steps to Test no longer say to set `config.inCluster: false`.

File: charts/headlamp/README.md (hunk 6 above)
Comment (cmcgalliard) [reply charts/headlamp/README.md]: Done — switched from the `KUBECONFIG` env var to the `-kubeconfig` argument. Since `config.setKubeConfigPath` returns early in in-cluster mode before reading the env var, passing the joined paths as `-kubeconfig` avoids that early return, so the in-cluster context and the mounted kubeconfig secrets now load together into the same store. No backend change needed.

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for these changes.


Can you please have a look at the git commits to see if they meet the contribution guidelines? We use a Linux kernel style of git commits. See the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) for general context, and please see previous git commits with git log for examples.

<details>
<summary>Commits that need attention</summary>

- `Update README to clarify kubeconfigSecrets usage` — Missing `area: description` prefix — e.g. `frontend: HomeButton: Fix so it navigates to home` or `backend: config: Add enable-dynamic-clusters flag`.

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for this PR.


A few of the commits don't quite follow the project guidelines. We use Linux kernel style for git commits — have a look at the [contributing guide](https://headlamp.dev/docs/latest/contributing#2-follow-commit-guidelines) and previous commits with `git log`.

<details>
<summary>Commits that need attention</summary>

- `Update README to clarify kubeconfigSecrets usage` — Missing `area: description` prefix — e.g. `frontend: HomeButton: Fix so it navigates to home` or `backend: config: Add enable-dynamic-clusters flag`.
- `Potential fix for pull request finding` — Missing `area: description` prefix — e.g. `frontend: HomeButton: Fix so it navigates to home` or `backend: config: Add enable-dynamic-clusters flag`.

</details>

<details>
<summary>Commit guidelines</summary>

- Use atomic commits focused on a single change.
- Use the title format `<area>: <Description of changes>` — description must start with a capital letter.
- Keep the title under 72 characters (soft requirement).
- Explain the intention and why the change is needed.
- Make commit titles meaningful and describe what changed.
- Do not add code that a later commit rewrites; squash or reorder commits instead.
- Do not include `Fixes #NN` in commit messages.

Good examples:

- `frontend: HomeButton: Fix so it navigates to home`
- `backend: config: Add enable-dynamic-clusters flag`

</details>

**Comment by cmcgalliard** [comment]
Cleaned up the commit history: squashed everything into a single atomic commit `charts: Add multi-cluster support using Kubernetes secrets` following the `<area>: <Description>` format. This also drops the two follow-up README commits (including the one that had a `Co-authored-by: Copilot Autofix` trailer, which was tripping EasyCLA) — their changes are folded into the single commit, which is authored solely by me.

**Comment by cmcgalliard** [comment]
@illume  - anything else you want to see before this gets merged?

**Comment by vtushar06** [comment]
@cmcgalliard I think this Pr is good to go, LGTM @illume

**Review by illume — CHANGES_REQUESTED** [review]
Thanks for working on this.


Would you mind addressing the open Copilot review comments? Please mark each comment as resolved after addressing it.
